### PR TITLE
Add a directory reader function

### PR DESCRIPTION
**Pull request by craig-cogdill**
I need something like this in the NICSetupAgent in order to scrape the /etc/sysconfig/network-scripts directory for valid interface config files.

### CODE REVIEW

File: test/ToolsTestFileIO.cpp [hunk 1]
```diff
@@ -954,6 +935,92 @@ TEST_F(TestFileIO, DISABLED_System_Performance_FileIO_DirectoryReader__vs_Boost_
...
+TEST_F(TestFileIO, CreateMultipleFilesAndReadThemIn) {
+   // Create subdirectory: /tmp/TempDirectoryTestOfFileIO/TestDir/
+   std::string newTestDirName = "TestDir";
+   std::string newDirectoryPath = mTestDirectory + std::string("/") + newTestDirName;
+   std::string createdDirectoryPath = CreateSubDirectory(newTestDirName, mTestDirectory);
+   EXPECT_TRUE(FileIO::DoesDirectoryExist(newDirectoryPath));
+   EXPECT_EQ(newDirectoryPath, createdDirectoryPath);
+
+   // Add multiple files
+   std::vector<std::string> filenames = {"test1", "test2", "test3", "test4"};
```
Comment (KjellKod) [inline test/ToolsTestFileIO.cpp]: Can you please merge the functionality to a function that takes a input and runs all the logic?  It seems that the functions are almost identical. 

Then the tests would be minimalistic and very easy to read
```
TEST_F(TestFileIO, CreateMultipleFilesAndReadThemIn) {
   auto filenames = std::vector<std::string>{...};
    CreateDirectoryContent(path, filenames);
   auto contentResult = FileIO::GetDirectoryContents(path);
   ASSERT_TRUE(contentREsult.hasSuccess());
   EXPECT_TRUE(VerifyDirectoryContent(filenames, contentResult.result);
}
```

File: test/ToolsTestFileIO.cpp [hunk 2]
```diff
@@ -954,6 +935,92 @@ TEST_F(TestFileIO, DISABLED_System_Performance_FileIO_DirectoryReader__vs_Boost_
...
+   for (const auto& filename : vecOfDirContents) {
+      EXPECT_TRUE(std::find(vecOfDirContents.begin(), vecOfDirContents.end(), filename) != vecOfDirContents.end());
+   }
+}
+
+TEST_F(TestFileIO, DirectoryDoesNotExist_NoFilesReturned) {
+   // Create subdirectory: /tmp/TempDirectoryTestOfFileIO/TestDir/
+   std::string newTestDirName = "TestDir";
+   std::string newDirectoryPath = mTestDirectory + std::string("/") + newTestDirName;
+   std::string createdDirectoryPath = CreateSubDirectory(newTestDirName, mTestDirectory);
```
Comment (KjellKod) [inline test/ToolsTestFileIO.cpp]: why do you create a directory that you are not using further down the test?

File: test/ToolsTestFileIO.cpp [hunk 3]
```diff
@@ -954,6 +935,92 @@ TEST_F(TestFileIO, DISABLED_System_Performance_FileIO_DirectoryReader__vs_Boost_
...
+
+   // Add 1 file
+   std::vector<std::string> filenames = {"test1", "test2", "test3", "test4"};
+   for (const auto& filename : filenames) {
+      std::string createdFilePath = CreateFile(createdDirectoryPath, filename);
+      EXPECT_TRUE(FileIO::DoesFileExist(createdFilePath));
+   }
+
+   // Read directory contents
+   auto dirContentsResult = FileIO::GetDirectoryContents("directory/does/not/exist");
```
Comment (KjellKod) [inline test/ToolsTestFileIO.cpp]: isn't the test these 4 lines and nothing else? 
I think it makes sense to remove everything above, don't you?

File: test/ToolsTestFileIO.cpp [hunk 4]
```diff
@@ -954,6 +935,92 @@ TEST_F(TestFileIO, DISABLED_System_Performance_FileIO_DirectoryReader__vs_Boost_
...
+   EXPECT_FALSE(dirContentsResult.HasSuccess());
+   auto vecOfDirContents = dirContentsResult.result; 
+   EXPECT_EQ(0, vecOfDirContents.size());
+}
+
+TEST_F(TestFileIO, EmptyDirectory_NoFilesReturnedButNoFailure) {
+   // Create subdirectory: /tmp/TempDirectoryTestOfFileIO/TestDir/
+   std::string newTestDirName = "TestDir";
+   std::string newDirectoryPath = mTestDirectory + std::string("/") + newTestDirName;
```
Comment (KjellKod) [inline test/ToolsTestFileIO.cpp]:
```
auto filenames = std::vector<std::string>{/*empty*/};
CreateDirectoryContent(path, filenames);
```
etc. 

I.e. this test and the other tests can follow the template oulined in the comments above

**Comment by KjellKod** [comment]
Awesome 🎉  🎈  function @craig-cogdill I can't believe we didn't have it earlier. 
Please ping me when I can review the tests again

File: test/ToolsTestFileIO.cpp [hunk 5]
```diff
@@ -954,6 +935,56 @@ TEST_F(TestFileIO, DISABLED_System_Performance_FileIO_DirectoryReader__vs_Boost_
...
+
+TEST_F(TestFileIO, DirectoryDoesNotExist_NoFilesReturned) {
+   auto dirContentsResult = FileIO::GetDirectoryContents("directory/does/not/exist");
+   EXPECT_FALSE(dirContentsResult.HasSuccess());
+   auto vecOfDirContents = dirContentsResult.result; 
+   EXPECT_EQ(0, vecOfDirContents.size());
+}
+
+TEST_F(TestFileIO, EmptyDirectory_NoFilesReturnedButNoFailure) {
+   std::vector<std::string> filenames = {};
```
Comment (KjellKod) [inline test/ToolsTestFileIO.cpp]: 💯 isn't this way way better?

**Comment by KjellKod** [comment]
👍

**Comment by KjellKod** [comment]
@craig-cogdill please hound someone to review this too @john-gress or @shwetabhandare maybe?

File: src/FileIO.cpp [hunk 6]
```diff
@@ -580,5 +580,24 @@ namespace FileIO {
...
       return (0 == rc);
    }
+
+   Result<std::vector<std::string>> GetDirectoryContents(const std::string& directory) {
+     std::vector<std::string> filesInDirectory;
+     DIR* dir;
+     struct dirent* entry;
+     if ((dir = opendir (directory.c_str())) != NULL) {
+       while ((entry = readdir (dir)) != NULL) {
+         if (!(std::string(entry->d_name) == "." || std::string(entry->d_name) == "..")) {
```
Comment (john-gress) [inline src/FileIO.cpp]: I think you would be better off using the `d_type == DT_REG`. That way you would only include regular files in the vector. And you might call this function, GetDirectoryRegularFiles.

Comment (shwetabhandare) [inline src/FileIO.cpp]: What about hidden files? Would this function add those to filesInDirectory?

File: test/ToolsTestFileIO.cpp [hunk 7]
```diff
@@ -954,6 +935,56 @@ TEST_F(TestFileIO, DISABLED_System_Performance_FileIO_DirectoryReader__vs_Boost_
...
+   EXPECT_TRUE(FileIO::DoesDirectoryExist(newDirectoryPath));
+   EXPECT_EQ(newDirectoryPath, createdDirectoryPath);
 
+   // Add files to directory
+   for (const auto& filename : filenamesToTouch) {
+      std::string createdFilePath = CreateFile(createdDirectoryPath, filename);
+      EXPECT_TRUE(FileIO::DoesFileExist(createdFilePath));
+   }
+   return createdDirectoryPath;
+}
```
Comment (shwetabhandare) [inline test/ToolsTestFileIO.cpp]: Can you add a test to have hidden files and see if they are added to the directory contents?

File: src/FileIO.cpp [hunk 8]
```diff
@@ -580,5 +580,24 @@ namespace FileIO {
 
       return (0 == rc);
    }
+
+   Result<std::vector<std::string>> GetDirectoryContents(const std::string& directory) {
```
Comment (shwetabhandare) [inline src/FileIO.cpp]: Is this recursive?

Comment (craig-cogdill) [inline src/FileIO.cpp]: No

Comment (shwetabhandare) [inline src/FileIO.cpp]: What would happen if there was a sub-directory ? Would that get appended to filesInDirectory as another file?

File: src/FileIO.cpp (hunk 6 above)
Comment (craig-cogdill) [inline src/FileIO.cpp]: @shwetabhandare Yes hidden files should be included. I will add that test below. Good suggestion

File: src/FileIO.cpp [hunk 9]
```diff
@@ -580,5 +580,25 @@ namespace FileIO {
...
+     std::vector<std::string> filesInDirectory;
+     DIR* dir;
+     struct dirent* entry;
+     if ((dir = opendir (directory.c_str())) != NULL) {
+       while ((entry = readdir (dir)) != NULL) {
+         // if (!(std::string(entry->d_name) == "." || std::string(entry->d_name) == "..")) {
```
Comment (john-gress) [inline src/FileIO.cpp]: Commented out code.

File: test/ToolsTestFileIO.cpp (hunk 7 above)
Comment (craig-cogdill) [inline test/ToolsTestFileIO.cpp]: Addressed on line 992

File: src/FileIO.cpp (hunk 8 above)
Comment (craig-cogdill) [inline src/FileIO.cpp]: It only looks for entries that are type DT_REG - so no directories should be pushed back.

**Comment by KjellKod** [comment]
@shwetabhandare can you take another look at this?

**Comment by shwetabhandare** [comment]
👍

**Pull request merged**